Use WP_Screen for script/styles loading.

jrfnl opened the pull request:
I might be missing something, but it looked like the scripts/styles were not being loaded on all possible back-end pages which could have CMB2 metaboxes.
Also, the previous array contained two invalid values.

Suggestion: better yet would be to use the `WP_Screen` object via `get_current_screen()` to check for the right pages.

jtsternberg commented:
Agree about using `WP_Screen` object. But, the enqueueing for the term/user pages is handled here: https://github.com/WebDevStudios/CMB2/blob/master/includes/CMB2_hookup.php#L260-L265

In my testing, the js/css is loaded in all the expected places, term pages, user profile, comment pages, post pages, etc.

jrfnl commented:
Ah, there we go. It did feel very much like I was missing something, so good.

Let me just change over this PR to using WP_Screen in that case, or would you rather receive a new PR for that ?

jtsternberg commented:
Let's do a new PR to keep the commit history a bit cleaner (unless you feel like scrubbing your branch history).

jrfnl commented:
I'll just rebase & squash, no worries.

jrfnl commented:
So this should only check for the edit comments and post pages ? Not term edit or user pages. Correct ?

jtsternberg commented:
Yah, basically the two contexts where `add_meta_box` is used, posts and comments. Need to be sure it works on new and edit pages.

jrfnl commented:
No worries, in WP_Screen they are not very different, new items just have the `add` `action`.

PR refreshed.

jtsternberg commented:
`parent_base

jrfnl commented:
@jtsternberg That's weird, it shouldn't be.
Hmm.. just looked into it deeper. Looks like WP does not have all information for WP_Screen setup on the `admin_enqueue_scripts` hook, which is weird as the `current_screen` action has already run by that time.

Let me play round with this a bit more to see which other info is available and usable for this situation.

jrfnl commented:
Another weird thing I'm running into - looks like the `do_scripts` function - in my test set up - is actually being called 8 times, even though it looks like you've tried to prevent that by having the `once` method. So I guess there's a bug in the `once` method or is this not what you were trying to accomplish there ?

If you like, I can fix that too - just a simple matter of wrapping the `add_filter` within `once` within a `has_filter` call.

jrfnl committed:
Refresh the conditional for script & style loading.

Change the script & style loading over to using the WP_Screen object for detecting the correct pages to load on.

jtsternberg commented:
Yah, that's odd. I think it probably has something to do w/ the serializing of the func args and the `$this` in the callback being unique. Anyway, I think it would be good to fix the `once` issue.. possibly w/ something like a `has_filter` proxy method. I'd be happy to see what you come up with.

jrfnl commented:
Ok, will do that in a separate PR.

For now, I've refreshed this PR with some code which *should* work on the `admin_enqueue_script` hook.
The scripts will load on add/edit pages, but not on the overview pages with the current code.

jrfnl commented:
> I think it probably has something to do w/ the serializing of the func args and the $this in the callback being unique. Anyway, I think it would be good to fix the once issue.. possibly w/ something like a has_filter proxy method.

You're right, it does have to do with the $this being different objects.
The oembed handler actions are fine as they use the same object via the `cmb2_ajax()` function and the `register_scripts` method is fine too with it being a static method.

It's just the calls from `post_hooks()` and `comment_hooks()` which pass the object which are troublesome.

tw2113 commented:
What's the status of this one @jrfnl ? Seeing any issues with it in your own dealings?

jrfnl commented:
Sorry haven't had a chance to work on this lately. Will get back to it at some point though ;-)
